### PR TITLE
Override HDUList.copy()

### LINKED ISSUE #7211: HDUList.copy() returns a list
Currently ``HDUList.copy()`` returns a list rather than an ``HDUList``:

```python
In [1]: from astropy.io.fits import HDUList

In [2]: hdulist = HDUList()

In [3]: hdulist.copy()
Out[3]: []

In [4]: type(_)
Out[4]: list
```

This is with Python 3.6.

### PR DESCRIPTION
**Pull request by ritiek**
Fixes #7211. Since `astropy.io.fitsHDUList` inherits from `List` object, `HDUList.copy()` returns a list. Overriding this `copy()` method should solve this problem.

### CODE REVIEW

**Comment by bsipocz** [comment]
@ritiek - a fits maintainer will do a proper review for this PR, but prior to that, please add a regression test (e.g. just the very same one that was reported in #7218)

**Comment by ritiek** [comment]
I've added a regression test. Let me know if there is anything else needs to be done. :)

File: astropy/io/fits/hdu/hdulist.py [hunk 1]
```diff
@@ -510,6 +510,10 @@ def fileinfo(self, index):
 
         return output
 
+    def copy(self):
+        # Return an HDUList when calling HDUList.copy() (issue #7211)
+        return self[:]
```
Comment (pllim) [inline astropy/io/fits/hdu/hdulist.py]: This returns a view, not a deep copy. It is unclear whether `copy()` should return a shallow or deep copy. I'll let the maintainers decide.

File: astropy/io/fits/tests/test_hdulist.py [hunk 2]
```diff
@@ -376,6 +376,15 @@ def test_file_like_3(self):
...
         assert fits.info(self.temp('tmpfile.fits'), output=False) == info
 
+    def test_copy_list(self):
+        """
+        Tests that HDUList.copy() returns an HDUList (regression test
+        for ticket:7211).
+        """
+        hdul = fits.HDUList()
+        new_hdul = hdul.copy()
+        assert isinstance(new_hdul, fits.HDUList)
```
Comment (pllim) [inline astropy/io/fits/tests/test_hdulist.py]: Should also test that the *content* is unchanged.

Comment (drdavella) [reply astropy/io/fits/tests/test_hdulist.py]: It would also be useful to do `assert new_hdul is not hdul` to make sure that we've actually gotten a copy.

File: astropy/io/fits/hdu/hdulist.py (hunk 1 above)
Comment (drdavella) [reply astropy/io/fits/hdu/hdulist.py]: It seems to me like the expected behavior would be a deep copy, meaning that the HDU objects in each of the lists should be equal but different. But @saimn should weigh in as well.

Comment (saimn) [reply astropy/io/fits/hdu/hdulist.py]: I think it should be a shallow copy, because `list.copy` and `copy.copy(hdulist)` do shallow copies, and also because it can be useful to copy just the container for instance to add a new HDU.

But there is also another discussion in #6395 about this and `copy.deepcopy` which do not work when a HDUList is read from a file (because by default it inherits `list` which tries to serialize the object).

File: astropy/io/fits/tests/test_hdulist.py (hunk 2 above)
Comment (ritiek) [reply astropy/io/fits/tests/test_hdulist.py]: @pllim Should we check that against an `HDUList` that actually contains `HDU` or just with an empty `HDUList`?

**Comment by ritiek** [comment]
@pllim I've added a change log in 8439594d5742167a3ca317e35e64899b9a0d605d.

File: astropy/io/fits/hdu/hdulist.py [hunk 3]
```diff
@@ -510,6 +510,10 @@ def fileinfo(self, index):
 
         return output
 
+    def copy(self):
+        # Return a shallow copy when calling HDUList.copy() (issue #7211)
```
Comment (saimn) [inline astropy/io/fits/hdu/hdulist.py]: This needs a real docstring instead of the comment (no need to put the issue number here).

File: astropy/io/fits/tests/test_hdulist.py (hunk 2 above)
Comment (saimn) [reply astropy/io/fits/tests/test_hdulist.py]: Your HDUList should contain HDUs (Primary + Image or BinTable) and check that it is (not) copied depending on the decision on the shallow copy.

Comment (ritiek) [reply astropy/io/fits/tests/test_hdulist.py]: Alright, I've updated the test. How does it look now?

File: astropy/io/fits/hdu/hdulist.py (hunk 3 above)
Comment (ritiek) [reply astropy/io/fits/hdu/hdulist.py]: OK, I've changed it to represent a docstring.

File: astropy/io/fits/hdu/hdulist.py [hunk 4]
```diff
@@ -510,6 +510,13 @@ def fileinfo(self, index):
 
         return output
 
+    def copy(self):
+        """
+        Return a shallow copy when calling HDUList.copy().
```
Comment (pllim) [inline astropy/io/fits/hdu/hdulist.py]: I don't think it should be a shallow copy, now that I found a similar discussion for FITS header in #4990.

Comment (pllim) [reply astropy/io/fits/hdu/hdulist.py]: @saimn , ah, I just saw your comment about shallow copy in my emails (they were hidden in the browser). If you think it is okay, then I am officially overruled. :smile:

**Comment by pllim** [comment]
I think the failure at https://travis-ci.org/astropy/astropy/jobs/344344771 will be fixed by @astrofrog 's segfault PR?

**Comment by saimn** [comment]
@pllim - yeah, discussing in hidden comments is sometimes hard to follow. 

Reading again #6395 and #4990, it seems the majority would prefer a deep copy by default. We should maybe support both options, with a keyword (`deep=True` ?), like in `Table.copy`. Then we could also use this in `__deepcopy__`.

@ritiek - Would you mind trying to implement this ? The deep copy solution is already drafted in https://github.com/astropy/astropy/issues/6395#issuecomment-318198524.

**Comment by saimn** [comment]
@pllim - about test failures, I think (and hope) so !

File: astropy/io/fits/tests/test_hdulist.py [hunk 5]
```diff
@@ -376,6 +376,24 @@ def test_file_like_3(self):
...
+        n = np.arange(10.0)
+        primary_hdu = fits.PrimaryHDU(n)
+        hdu = fits.ImageHDU(n)
+
+        hdul = fits.HDUList([primary_hdu, hdu])
+        new_hdul = hdul.copy()
+        new_hdul[1] = primary_hdu
+
+        assert isinstance(new_hdul, fits.HDUList)
+        assert not new_hdul[1] == hdul[1]
```
Comment (drdavella) [inline astropy/io/fits/tests/test_hdulist.py]: If you're doing this check, it probably also makes sense to do `assert new_hdul[0] == hdul[0]` as well.

Comment (drdavella) [reply astropy/io/fits/tests/test_hdulist.py]: Although to be honest I'm not sure if checking for equality between `HDU` objects is that informative anyway.

Comment (ritiek) [reply astropy/io/fits/tests/test_hdulist.py]: I thought so too. I think a better option would be to use [`id()`](https://docs.python.org/3/library/functions.html#id). How about something like this:
```python
>>> hdul = fits.HDUList([primary_hdu, hdu])
# shallow copy
>>> new_hdul = hdul.copy(deep=False)
# different HDULists
>>> assert not (id(new_hdul) == id(hdul))
# share same item
>>> assert id(new_hdul[0]) == id(hdul[0])
```

Comment (drdavella) [reply astropy/io/fits/tests/test_hdulist.py]: @ritiek in the case of a shallow copy, it should be possible just to use `is` as the comparator, which should be the same as testing for equivalent `id`s. For the deep copy case, it will be necessary to test object equality, although I'm not whether HDU objects have a meaningful definition of equality.

Comment (ritiek) [reply astropy/io/fits/tests/test_hdulist.py]: @drdavella Ok, got it. So, this is what we have (not sure about equating HDU objects either):
```python
>>> shallow_copy = hdul.copy(deep=False)
>>> deep_copy = hdul.copy(deep=True)

>>> assert isinstance(shallow_copy, fits.HDUList)
>>> assert isinstance(deep_copy, fits.HDUList)

>>> assert not (id(shallow_copy) == id(hdul))
>>> assert shallow_copy[0] is hdul[0]
>>> assert shallow_copy[1] is hdul[1]

>>> assert not (id(deep_copy) == id(hdul))
>>> assert not deep_copy[0] == hdul[0]
>>> assert not deep_copy[1] == hdul[1]
```
I wanted to know whether this looks good before I update this test downstream (and this conversation becomes hard to follow up :3).

Comment (drdavella) [reply astropy/io/fits/tests/test_hdulist.py]: @ritiek this is getting there, but it's still not quite right. Here's what I think it should be:
```python
>>> shallow_copy = hdul.copy(deep=False)
>>> deep_copy = hdul.copy(deep=True)

>>> assert isinstance(shallow_copy, fits.HDUList)
>>> assert isinstance(deep_copy, fits.HDUList)

>>> assert shallow_copy is not hdul
>>> assert shallow_copy[0] is hdul[0]
>>> assert shallow_copy[1] is hdul[1]

>>> assert deep_copy is not hdul
>>> assert deep_copy[0] is not hdul[0]
>>> assert deep_copy[1] is not hdul[1]
```

In an ideal world, we'd also be able to do the following:

```python
>>> assert deep_copy[0] == hdul[0]
>>> assert deep_copy[1] == hdul[1]
```

However, since we don't have a good definition of equality for HDUs at the moment, I don't think that will work. It **may** be possible to do this instead:

```python
>>> assert deep_copy[0].header == hdul[0].header
>>> assert deep_copy[1].header == hdul[1].header
>>> (assert deep_copy[0].data == hdul[0].data).all()
>>> (assert deep_copy[1].data == hdul[1].data).all()
```

But that may be overkill.

**Comment by ritiek** [comment]
`HDUList.copy()` now returns a deep copy as defaults and added some tests for it as well.

File: CHANGES.rst [hunk 6]
```diff
@@ -187,6 +187,9 @@ astropy.io.misc
 astropy.io.fits
 ^^^^^^^^^^^^^^^
 
+- Override ``HDUList.copy()`` to return a copy instead
+  of ``list``. [#7218]
```
Comment (saimn) [inline CHANGES.rst]: I would suggest something like *"... to return a HDUList instance, and do a deep copy by default"*.

File: astropy/io/fits/hdu/hdulist.py [hunk 7]
```diff
@@ -510,6 +510,17 @@ def fileinfo(self, index):
 
         return output
 
+    def copy(self, deep=True):
+        """
+        Return a copy when calling HDUList.copy().
```
Comment (saimn) [inline astropy/io/fits/hdu/hdulist.py]: Needs documentation for `deep`.

**Comment by saimn** [comment]
This is both a bugfix (returning an HDUList instance) and a new feature (deep copy by default with a new keyword). So I'm not sure about the milestone. Maybe we should go for 3.1, thoughts @drdavella @MSeifert04  ?

File: CHANGES.rst (hunk 6 above)
Comment (ritiek) [reply CHANGES.rst]: OK, I've addressed your review comments.

**Comment by pllim** [comment]
Given that fixing anything in FITS might break something obscure for some people, I vote for 3.1, so at least people can stick with 3.0.x if this does not work for them... :grimacing:

**Comment by MSeifert04** [comment]
I would have preferred not having the deep argument and just do the shallow copy. One should implement the `__deepcopy__` special method if one wants to support deep copies. It's rather odd to deviate that much from the superclass and previous astropy-releases (which just did the shallow copy).

**Comment by saimn** [comment]
@MSeifert04 - I was also in favor of shallow copy but in #6395 most people were expecting deep copy I think (and `Table.copy` is deep by default, I guess because people expect deep copies most of the time). Also in https://github.com/astropy/astropy/issues/6395#issuecomment-318198524 you were proposing a deep copy implementation ;)

 @ritiek - I set the milestone to 3.1, could you just squash your commits into one ?

**Comment by ritiek** [comment]
@saimn Yep, squashed them!

**Comment by saimn** [comment]
I agree with you @MSeifert04 but it seems (from the discussion in the other issues) that people mostly expect a deepcopy by default, and `Table.copy` is also deep by default. So I don't know, I have no strong feelings - both options are acceptable, but we should make a decision...
@drdavella - What's your opinion?

**Comment by drdavella** [comment]
@saimn I agree that both options are acceptable. Maybe the best approach is in fact to implement `__copy__` and `__deepcopy__` and then to have the `copy` method simply be a shallow copy, as @MSeifert04 suggests. This also might let us milestone earlier than 3.1 if we want since there's no API breakage.

**Comment by saimn** [comment]
@drdavella - Thanks, that works for me! 

@ritiek - Could you update the PR with @drdavella's solution ? Hopefully this should be the final version ;)

**Comment by ritiek** [comment]
@saimn Okay, I'll hopefully add `__copy__()` and `__deepcopy__()` methods this evening.

File: astropy/io/fits/hdu/hdulist.py [hunk 8]
```diff
@@ -510,6 +510,46 @@ def fileinfo(self, index):
...
+        Returns
+        -------
+        self : HDUList
+            A shallow copy of this :class:`HDUList` object.
+
+        """
+
+        return self[:]
+
+    def copy(self):
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/hdulist.py]: You could even use `copy = __copy__` and avoid the layer of indirection.

Comment (ritiek) [reply astropy/io/fits/hdu/hdulist.py]: Wow! Thanks. I had no idea it was possible to something like this for a class method.

Comment (ritiek) [reply astropy/io/fits/hdu/hdulist.py]: Wow! Thanks. I had no idea it was possible to do something like this for a class method.

File: astropy/io/fits/tests/test_hdulist.py [hunk 9]
```diff
@@ -376,6 +376,52 @@ def test_file_like_3(self):
...
+        shallow copy (regression test for ticket:7211).
+        """
+
+        n = np.arange(10.0)
+        primary_hdu = fits.PrimaryHDU(n)
+        hdu = fits.ImageHDU(n)
+        hdul = fits.HDUList([primary_hdu, hdu])
+
+        shallow_copy = hdul.copy()
+        magic_shallow_copy = hdul.__copy__()
```
Comment (saimn) [inline astropy/io/fits/tests/test_hdulist.py]: You should test with `copy.copy` (and `copy.deepcopy` below) instead of using the magic methods.

File: astropy/io/fits/tests/test_hdulist.py [hunk 10]
```diff
@@ -376,6 +376,52 @@ def test_file_like_3(self):
...
+
+        assert isinstance(shallow_copy, fits.HDUList)
+        assert isinstance(magic_shallow_copy, fits.HDUList)
```
Comment (saimn) [inline astropy/io/fits/tests/test_hdulist.py]: Maybe you could loop on the two hdu copies to avoid repeating the test logic (e.g. `for hducopy in (hdul.copy(), copy.copy(hdul))`).

File: CHANGES.rst [hunk 11]
```diff
@@ -187,6 +187,10 @@ astropy.io.misc
 astropy.io.fits
 ^^^^^^^^^^^^^^^
 
+  Create magic methods ``__copy__()`` and ``__deepcopy__`` on an ``HDUList``
```
Comment (saimn) [inline CHANGES.rst]: Missing a list bullet, and I think that parenthesis are unnecessary for method names (or at least it should consistent).

**Comment by saimn** [comment]
@ritiek - could you also rebase on master, this may fix the circleci build.

File: astropy/io/fits/tests/test_hdulist.py (hunk 9 above)
Comment (ritiek) [reply astropy/io/fits/tests/test_hdulist.py]: I am not sure what you mean by "test with `copy.copy`". Why do we need to use `copy` module here instead of testing the methods we added?

File: CHANGES.rst (hunk 11 above)
Comment (ritiek) [reply CHANGES.rst]: Ah, I see. Gotta fix that.

**Comment by ritiek** [comment]
I've rebased this PR on master branch but for some Travis reason, all tests seem to passing [except for this one](https://travis-ci.org/astropy/astropy/jobs/351151907) (network timed out).

`copy.deepcopy` indeed gave problems, so I've added an optional `memo=None` parameter to `__deepcopy__()` and it works good now. I am not sure if and what kind of documentation should we add about `memo` parameter?

**Comment by MSeifert04** [comment]
I dont think we need to/should document special methods at all.

**Comment by MSeifert04** [comment]
Ah, sorry. My comment was only with regard to `__deepcopy__`. The `copy` method is public so it should be documented. Could you restore the `__copy__` docstring? 😅

**Comment by ritiek** [comment]
My bad, I thought you meant for both those methods. Anyway, I've restored docstring for `__copy__`. :D

File: astropy/io/fits/tests/test_hdulist.py [hunk 12]
```diff
@@ -376,6 +377,43 @@ def test_file_like_3(self):
...
+        primary_hdu = fits.PrimaryHDU(n)
+        hdu = fits.ImageHDU(n)
+        hdul = fits.HDUList([primary_hdu, hdu])
+
+        hdulcopy = copy.deepcopy(hdul)
+
+        assert isinstance(hdulcopy, fits.HDUList)
+        assert hdulcopy is not hdul
+
+        for index in range(1):
```
Comment (drdavella) [inline astropy/io/fits/tests/test_hdulist.py]: Shouldn't this be `range(2)`?

Comment (ritiek) [reply astropy/io/fits/tests/test_hdulist.py]: Ah, yes. Silly me. Fixed it.

File: CHANGES.rst [hunk 13]
```diff
@@ -202,6 +202,10 @@ astropy.io.misc
 astropy.io.fits
 ^^^^^^^^^^^^^^^
 
+- Create magic methods ``__copy__`` and ``__deepcopy__`` on an ``HDUList``
+  class instance [#7218].
+- Override ``HDUList.copy()`` to return a shallow HDUList instance [#7218].
```
Comment (MSeifert04) [inline CHANGES.rst]: Please leave a line between the two points

File: CHANGES.rst [hunk 14]
```diff
@@ -202,6 +202,10 @@ astropy.io.misc
 astropy.io.fits
 ^^^^^^^^^^^^^^^
 
+- Create magic methods ``__copy__`` and ``__deepcopy__`` on an ``HDUList``
```
Comment (MSeifert04) [inline CHANGES.rst]: Normally this is called "added support for `copy.copy` and `copy.deepcopy` for `HDUList`". No need to mention the magic method names in the changelog.

File: CHANGES.rst [hunk 15]
```diff
@@ -202,6 +202,10 @@ astropy.io.misc
...
+  class instance [#7218].
```
Comment (MSeifert04) [inline CHANGES.rst]: Minor nitpick: The dot should be in front of the issue number `class instance. [#7218]`

File: CHANGES.rst (hunk 13 above)
Comment (MSeifert04) [reply CHANGES.rst]: Please put the dot in front of the issue number.

File: astropy/io/fits/hdu/hdulist.py [hunk 16]
```diff
@@ -510,6 +510,25 @@ def fileinfo(self, index):
 
         return output
 
+    def __copy__(self):
+        """
+        Return a shallow copy of the object.
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/hdulist.py]: `object` -> `HDUList`

File: astropy/io/fits/hdu/hdulist.py [hunk 17]
```diff
@@ -510,6 +510,25 @@ def fileinfo(self, index):
...
+
+        Returns
+        -------
+        self : HDUList
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/hdulist.py]: shouldn't be `self` because it doesn't return `self`. What about `copy : HDUList`? Also please put single backticks around `HDUList`.

File: astropy/io/fits/hdu/hdulist.py [hunk 18]
```diff
@@ -510,6 +510,25 @@ def fileinfo(self, index):
...
+            A shallow copy of this :class:`HDUList` object.
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/hdulist.py]: The `:class:` is (probably) not needed. We do the links automagically.

File: astropy/io/fits/tests/test_hdulist.py [hunk 19]
```diff
@@ -376,6 +377,43 @@ def test_file_like_3(self):
         info = [(0, 'PRIMARY', 1, 'PrimaryHDU', 5, (100,), 'int32', '')]
         assert fits.info(self.temp('tmpfile.fits'), output=False) == info
 
+    def test_shallow_copy(self):
+        """
+        Tests that `HDUList.__copy__()`and `HDUList.copy()` returns a
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_hdulist.py]: `returns` -> `return`

File: astropy/io/fits/tests/test_hdulist.py [hunk 20]
```diff
@@ -376,6 +377,43 @@ def test_file_like_3(self):
...
+        shallow copy (regression test for ticket:7211).
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_hdulist.py]: just use `#7211` there instead of the `ticket:7211`.

File: astropy/io/fits/tests/test_hdulist.py [hunk 21]
```diff
@@ -376,6 +377,43 @@ def test_file_like_3(self):
...
+
+        hdulcopy = copy.deepcopy(hdul)
+
+        assert isinstance(hdulcopy, fits.HDUList)
+        assert hdulcopy is not hdul
+
+        for index in range(len(hdul)):
+            assert hdulcopy[index] is not hdul[index]
+            assert hdulcopy[index].header == hdul[index].header
+            assert (hdulcopy[index].data == hdul[index].data).all()
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_hdulist.py]: That should be `np.assert_array_equal(hdulcopy[index].data, hdul[index].data)`. That also checks some other stuff like equal shapes, etc.

**Comment by saimn** [comment]
Merging since everybody approved, thanks @ritiek !
